File: core/src/ru/myitschool/dcrawler/entities/skills/JaggedSword.java
package ru.myitschool.dcrawler.entities.skills;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.effects.BloodiedEffect;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.DiceAction;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.math.NumberAction;
import ru.myitschool.dcrawler.entities.skills.action.SimpleCombinedAction;
import ru.myitschool.dcrawler.entities.skills.patterns.EntityTargetPattern;
import ru.myitschool.dcrawler.entities.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.entities.skills.targeting.EntityDisplayer;
import ru.myitschool.dcrawler.entities.skills.targeting.TargetDisplayer;

/**
 * Created by Voyager on 18.11.2017.
 */
public class JaggedSword extends Skill {

    MathAction attackAction = new DiceAction(6);

    int normalDamage = 3;
    int criticalDamage = 5;

    int rounds = 3;

    public JaggedSword(Entity doer) {
        super(doer);
        addPlayContainer().getEnemyPlay().addAction(new SimpleCombinedAction(this, attackAction) {
            int bleedDamage = 0;

            @Override
            protected void successEffect(Target target, int damage, FloatingDamageMark mark) {
                bleedDamage = normalDamage;
            }

            @Override
            protected void critSuccessEffect(Target target, int damage, FloatingDamageMark mark) {
                bleedDamage = criticalDamage;
            }

            @Override
            protected void afterEffect(Target target, int damage, FloatingDamageMark mark) {
                if (bleedDamage > 0) {
                    Entity entity = target.getEntity();
                    entity.addEffect(new BloodiedEffect(entity, new NumberAction(bleedDamage), rounds));
                }
                if (bleedDamage == normalDamage) {
                    mark.addText("Success");
                } else if (bleedDamage == criticalDamage) {
                    mark.addText("Critical success");
                }
            }
        });
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new EntityDisplayer());
    }

    @Override
    public TargetPattern getPattern() {
        return new EntityTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("blood.png");
    }

    @Override
    public String getName() {
        return "Jagged sword";
    }

    @Override
    public String getDescription() {
        return "Deal some damage and adds bleed to target";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 1;
    }

    @Override
    public int getDistanceMin() {
        return 1;
    }

    @Override
    public int getDistanceMax() {
        return 1;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 4;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_COOLDOWN;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_ENTITY;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return true;
    }

    @Override
    public boolean isWallTargets() {
        return false;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/JaggedSword.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.effects.BloodiedEffect
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.action.SimpleCombinedAction
import com.wolg_vlad.dcrawler.entities.skills.patterns.EntityTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.EntityDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer
import com.wolg_vlad.dcrawler.math.DiceAction
import com.wolg_vlad.dcrawler.math.MathAction
import com.wolg_vlad.dcrawler.math.NumberAction

/**
 * Created by Voyager on 18.11.2017.
 */
class JaggedSword(doer: Entity) : Skill(doer) {
    var attackAction: MathAction = DiceAction(6)
    var normalDamage = 3
    var criticalDamage = 5
    var rounds = 3

    override val displayers: Array<TargetDisplayer> = arrayOf(EntityDisplayer())
    override val pattern: TargetPattern = EntityTargetPattern(this)
    override val icon: Texture = Texture("blood.png")
    override val name: String = "Jagged sword"
    override val description: String = "Deal some damage and adds bleed to target"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 1
    override val distanceMin: Int = 1
    override val distanceMax: Int = 1
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 4
    override val type: Int = SKILL_TYPE_COOLDOWN
    override val targetType: Int = SKILL_TARGET_TYPE_ENTITY
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = true
    override val isWallTargets: Boolean = false

    init {
        addPlayContainer().enemyPlay.addAction(object : SimpleCombinedAction(this, attackAction) {
            var bleedDamage = 0
            override fun successEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                bleedDamage = normalDamage
            }

            override fun critSuccessEffect(target: Target?, damage: Int, mark: FloatingDamageMark?) {
                bleedDamage = criticalDamage
            }

            override fun afterEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                if (bleedDamage > 0) {
                    val entity = target.entity
                    entity!!.addEffect(BloodiedEffect(entity, NumberAction(bleedDamage), rounds))
                }
                if (bleedDamage == normalDamage) {
                    mark.addText("Success")
                } else if (bleedDamage == criticalDamage) {
                    mark.addText("Critical success")
                }
            }
        })
    }
}